base/optimized_generics.jl

```diff
@@ -0,0 +1,58 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module OptimizedGenerics
+
+# This file defines interfaces that are recognized and optimized by the compiler
+# They are intended to be used by data structure implementations that wish to
+# opt into some level of compiler optimizations. These interfaces are
+# EXPERIMENTAL and currently intended for use by Base only. They are subject
+# to change or removal without notice. It is undefined behavior to add methods
+# to these generics that do not conform to the specified interface.
+#
+# The intended way to use these generics is that data structures will provide
+# appropriate implementations for a generic. In the absence of compiler
+# optimizations, these behave like regular methods. However, the compiler is
+# semantically allowed to perform certain structural optimizations on
+# appropriate combinations of these intrinsics without proving correctness.
+
+# Compiler-recognized generics for immutable key-value stores (dicts, etc.)
+"""
+    module KeyValue
+
+Implements a key-value like interface where the compiler has liberty to perform
+the following transformations. The core optimization semantically allowed for
+the compiler is:
+
+    get(set(x, key, val), key) -> val
+
+where the compiler will recursively look through `x`. Keys are compared by
```

Comment:
> where the compiler will recursively look through `x`

What does this mean?

Reply:
`get(set(set(:x, 1), :y, 2), :x)` is foldable to `1`.

Reply:
That reasoning seems to rely on proving `!isequal(x,y)` in addition to non-egality.

For example:
```julia
julia> make_nan(x::Integer) = reinterpret(Float64, 0xFFF8_0000_0000_0000 | UInt32(x))

julia> make_nan(1) === make_nan(2)
false

julia> isequal(make_nan(1), make_nan(2))
true

julia> Base.PersistentDict(Base.PersistentDict(make_nan(1) => 2.0), make_nan(2) => 1.0)
Base.PersistentDict{Float64, Float64} with 1 entry:
  NaN => 1.0
```